src/Shout/SHRBTextStyler.class.st

```diff
@@ -836,6 +836,10 @@ SHRBTextStyler >> style: aText ast: ast [
 	parentheseLevel := 0.
 	braceLevel := 0.
 	self visitNode: ast.
+	"Second pass to ensure that all syntax errors are visible"
+	ast allErrorNotices do: [ :notice | 
+		notice isSyntaxError ifTrue: [
+			self addStyle: #invalid from: notice position to: notice position ] ].
```

Comment:
With this, do we need to style the other invalid positions or could this deal with all invalid things at once?
For example can we remove visitParseErrorNode: or not?

Reply:
This one adds red to a single character that might be outside the span of the node. Especially for 0-width errors where all positions are necessary outside.
It is also used by a few bad number tokens that have a span and a error position outside.
The visitParseErrorNode take care of styling the full node, not only the error position.

I didn't change that much the way Pharo report errors. I changed most of the API and the implementation but I did not touch a lot the error messages or their positions.
Some might feel weird because they are mechanical errors "I did not expect this character now in the stream" where the error position is where you are in the stream and the error is reported with `blabla->` pointing the unwanted character and stopping the parsing.

The parser does not work this way anymore and always continue in recovery mode to give you an AST, so the unwanted character might be consumed in the error node, but it might instead be used for something else meaningful in the AST. Leaving the error character outside the error node.
The question is therefore how to report such error visually, especially if the error node is empty?
I simply choose to force the highlighting of the character at the error position, because whatever the character might be reinserted to, it is still a culprit for the parsing law.

Note: This is consistent with how the styler worked previously with ad hoc heuristic (for instance it considered that the error position is always the first character that follows its span). I just make it correct (use real information available in the AST, not heuristics) and systematic.